src/bucket/BucketSnapshot.cpp

```diff
@@ -8,66 +8,77 @@
 #include "ledger/LedgerTxn.h"
 #include "ledger/LedgerTypeUtils.h"
 #include "util/XDRStream.h"
+#include <type_traits>
 
 namespace stellar
 {
-BucketSnapshot::BucketSnapshot(std::shared_ptr<Bucket const> const b)
+template <class BucketT>
+BucketSnapshotBase<BucketT>::BucketSnapshotBase(
+    std::shared_ptr<BucketT const> const b)
     : mBucket(b)
 {
     releaseAssert(mBucket);
 }
 
-BucketSnapshot::BucketSnapshot(BucketSnapshot const& b)
+template <class BucketT>
+BucketSnapshotBase<BucketT>::BucketSnapshotBase(
+    BucketSnapshotBase<BucketT> const& b)
     : mBucket(b.mBucket), mStream(nullptr)
 {
     releaseAssert(mBucket);
 }
 
+template <class BucketT>
 bool
-BucketSnapshot::isEmpty() const
+BucketSnapshotBase<BucketT>::isEmpty() const
 {
     releaseAssert(mBucket);
     return mBucket->isEmpty();
 }
 
-std::pair<std::optional<BucketEntry>, bool>
-BucketSnapshot::getEntryAtOffset(LedgerKey const& k, std::streamoff pos,
-                                 size_t pageSize) const
+template <class BucketT>
+std::pair<std::shared_ptr<typename BucketSnapshotBase<BucketT>::BucketEntryT>,
+          bool>
+BucketSnapshotBase<BucketT>::getEntryAtOffset(LedgerKey const& k,
+                                              std::streamoff pos,
+                                              size_t pageSize) const
 {
     ZoneScoped;
     if (isEmpty())
     {
-        return {std::nullopt, false};
+        return {nullptr, false};
```

Comment:
This is to avoid an extra copy. Previously, `SearchableBucketListSnapshot::load` was the high-level entry point for point loads and returned `std::shared_ptr<LedgerEntry>`. This lower level function, `getEntryAtOffset`, returned a `BucketEntry`, and we needed to construct a `LedgerEntry` anyway, so returning optional was fine.

However, now the Hot Archive BucketList returns a `HotArchiveBucketEntry` type. If we kept this lower level interface as an optional, we'd have to make a copy of the optional contents to construct the shared_ptr of HotArchiveBucketEntry returned by the higher level interface.

Alternatively, we could switch the higher level interfaces of `SearchableLiveBucketListSnapshot::load` and `SearchableHotArchiveBucketListSnapshot::load` to return optionals instead of shared_ptr, but that breaks some of the ltx cache that relies on shared_ptr.

Reply:
Hmm, I'm not 100% clear on this but it's not a big deal.